hicommonwealth/commonwealth-oss
# Updating democracy module for Substrate PR #5294.

## Description
Substrate PR #5294 (https://github.com/paritytech/substrate/pull/5294) brings many updates to the democracy module, which this PR aims to support.

Implementation TODO:
- [x] Upgrade API for latest democracy support
- [x] No more errors on democracy viewing
- [x] Fix existing democracy transactions
- [ ] ~Support proxy system queries~
- [ ] ~Support proxy system transactions~
- [x] Support adminstrative transactions (`reap_vote`, `unlock`)
- [ ] ~Review and make changes for `Instant` as needed~
- [x] Fix the locking mechanism to prevent proposal creation errors
  - I fixed this and liquid/illiquid seems accurate to the chain, but also sometimes doesn't make sense to me (edgeware votes locking even if you select 0.1x?)
- [x] Writing methods for tx fee calculation (Substrate now simulates txs and has floating gas prices).
- [x] Removing transactionfee because it no longer exists
- [x] Figure out why `remark` always fails to execute.
  - `remark` always fails, but `balance.setBalance` works, so I'm considering this a success.
- [x] Ensure conviction still works as expected.

UI fixes needed:
- [ ] Passing/failing text on proposals page looks weird.
- [ ] Some sort of notification on error, rather than just console printing?
- [ ] Some way to unlock funds (`unlock` call, on Kusama), or else they'll remain illiquid forever.
- [ ] Consider allowing unvote on Kusama. We may need to continue displaying proposals even after completion to allow for unvoting.
- [ ] Add input for custom balance when voting.
- [ ] Add visual display for passed-but-not-yet-dispatched proposal state (enactment time), and visual display for cooloff time too, if we want it.

## Motivation and Context
Kusama runs off a more bleeding-edge version of Substrate than Edgeware, and we want to support Kusama, so this set of changes is designed to get Kusama back online.

## How Has This Been Tested?
Code builds, and basic view tests. QA on local Kusama and local Edgeware chains. Query-only QA on live Kusama and live Edgeware chains.

Replication steps (Kusama):
1. Check out Kusama at Substrate commit hash 2eb9c260a.
2. Apply the following diff to decrease Democracy launch, voting, enactment, and cooloff times to minutes rather than days.:

```
diff --git a/bin/node/runtime/src/lib.rs b/bin/node/runtime/src/lib.rs
index ccd2eba78..9e02db39a 100644
--- a/bin/node/runtime/src/lib.rs
+++ b/bin/node/runtime/src/lib.rs
@@ -301,13 +301,13 @@ impl pallet_staking::Trait for Runtime {
 }
 
 parameter_types! {
-       pub const LaunchPeriod: BlockNumber = 28 * 24 * 60 * MINUTES;
-       pub const VotingPeriod: BlockNumber = 28 * 24 * 60 * MINUTES;
+       pub const LaunchPeriod: BlockNumber = 2 * MINUTES;
+       pub const VotingPeriod: BlockNumber = 2 * MINUTES;
        pub const FastTrackVotingPeriod: BlockNumber = 3 * 24 * 60 * MINUTES;
        pub const InstantAllowed: bool = true;
        pub const MinimumDeposit: Balance = 100 * DOLLARS;
-       pub const EnactmentPeriod: BlockNumber = 30 * 24 * 60 * MINUTES;
-       pub const CooloffPeriod: BlockNumber = 28 * 24 * 60 * MINUTES;
+       pub const EnactmentPeriod: BlockNumber = 1 * MINUTES;
+       pub const CooloffPeriod: BlockNumber = 1 * MINUTES;
        // One cent: $10,000 / MB
        pub const PreimageByteDeposit: Balance = 1 * CENTS;
 }
```

3. Build and run chain with `./target/release/substrate --alice --chain=dev --force-authoring --validator`.
4. Replace `edgeware-local` NodeInfo with `kusama-local` in the database (either manually or by resetting).
5. Load `http://localhost:8080/kusama-local/`. Verify the console prints `KUSAMA started.`.
6. Follow flow to link address "Alice" (`5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY`) via subkey (use "//Alice" on command line).
7. Create a democracy proposal as Alice. I typically use "remark". Use the private key to sign.
8. (Optional) Note the preimage as well.
9. Wait until the proposal becomes a referendum. Try to vote.
10. Ensure the proposal outcome is as expected.
11. At this point, if you try to repeat the process from step 7, you'll notice that all your funds are locked. This is where `unvote` and setting a vote quantity become important. Also, if you chose to register the preimage and then pass the proposal, you'll note in the execution event that the function call fails. The reason for this remains unclear to me.

Repeat this process with a local Edgeware node rather than Kusama to verify the changes are non-breaking.

## Types of changes

## Checklist:
